### PR TITLE
Onboarding

### PR DESCRIPTION
**Pull request by annybs**
This PR includes all basic device commands covering the full onboarding journey:

- Add device to network (initialize & assign to stake)
- Show device and stake info
- Start/restart/stop node corresponding to stake type - host, gateway, stargate
- Show status of node, i.e. whether it is running or not

There is also 'offboarding' capability via `device remove` to remove the device from the network:

- Unassign stake
- Stop node (if running)
- Delete device info

This closes multiple issues, which will link back to this PR.

### CODE REVIEW

**Comment by annybs** [comment]
This PR is a draft at the moment because there are multiple UX issues, as I will discuss on #19 

However, the basic functions are present and correct and available to check over now if desired.

File: .build/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,24 @@
+FROM node:14
```
Comment (adamkdean) [inline .build/Dockerfile:1]: Insert above L1:

```
# Copyright (C) 2021 Edge Network Technologies Limited
# Use of this source code is governed by a GNU GPL-style license
# that can be found in the LICENSE.md file. All rights reserved.
```

File: .build/auto_target.sh [hunk 2]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
```
Comment (adamkdean) [inline .build/auto_target.sh:1]: Insert after hashbang, like so:

```
#!/bin/bash
# Copyright (C) 2021 Edge Network Technologies Limited
# Use of this source code is governed by a GNU GPL-style license
# that can be found in the LICENSE.md file. All rights reserved.

set -e
```

File: .build/build.sh [hunk 3]
```diff
@@ -0,0 +1,29 @@
+#/bin/bash
```
Comment (adamkdean) [inline .build/build.sh:1]: Invalid hashbang. Also, insert comment, leaving line after, like so:

```
#!/bin/bash
# Copyright (C) 2021 Edge Network Technologies Limited
# Use of this source code is governed by a GNU GPL-style license
# that can be found in the LICENSE.md file. All rights reserved.

set -e
```

File: package.json [hunk 4]
```diff
@@ -24,23 +26,29 @@
...
-    "pkg": "^5.3.1",
     "ts-node": "^10.2.1",
     "typescript": "^4.4.2"
   },
   "dependencies": {
     "@edge/index-utils": "^0.1.0",
     "@edge/wallet-utils": "^0.14.2",
-    "@edge/xe-utils": "^1.0.1",
+    "@edge/xe-utils": "^1.0.4",
+    "@types/tar-stream": "^2.2.2",
```
Comment (adamkdean) [inline package.json:43]: `@types/tar-stream` should be dev dep

File: src/device/cli.ts [hunk 5]
```diff
@@ -2,59 +2,543 @@
...
+}
+
+const addHelp = (network: Network) => [
+  '\n',
+  'This command will add this device to the network, allowing it to operate as a node.\n\n',
+  'Adding a device will:\n',
+  '  - Initialize its identity if needed\n',
+  '  - Assign it to a stake\n\n',
+  'Stake assignment requires a blockchain transaction. After the transaction has been processed, this device can ',
+  'run a node of the staked type.\n\n',
```
Comment (adamkdean) [inline src/device/cli.ts:162]: staked type -> stake type? 

Not sure on this. Thoughts?

File: src/device/cli.ts [hunk 6]
```diff
@@ -2,59 +2,543 @@
...
+    const onChainWallet = await api.walletWithNextNonce(wallet.address)
+
+    const tx = xe.tx.sign({
+      timestamp: Date.now(),
+      sender: wallet.address,
+      recipient: wallet.address,
+      amount: 0,
+      data: {
+        action: 'unassign_device',
+        memo: 'Unassign device from stake',
```
Comment (adamkdean) [inline src/device/cli.ts:263]: As per brand guidelines, this should be title case. I recommend the memo `Unassign Device`.

File: src/device/cli.ts [hunk 7]
```diff
@@ -2,59 +2,543 @@
...
+
+  const tx = xe.tx.sign({
+    timestamp: Date.now(),
+    sender: wallet.address,
+    recipient: wallet.address,
+    amount: 0,
+    data: {
+      action: 'assign_device',
+      device: device.address,
+      memo: 'Assign device to stake',
```
Comment (adamkdean) [inline src/device/cli.ts:136]: Change to `Assign Device`

File: src/device/data.ts [hunk 8]
```diff
@@ -0,0 +1,200 @@
+import * as xe from '@edge/xe-utils'
```
Comment (adamkdean) [inline src/device/data.ts:1]: Prepend with 

```
// Copyright (C) 2021 Edge Network Technologies Limited
// Use of this source code is governed by a GNU GPL-style license
// that can be found in the LICENSE.md file. All rights reserved.
```

Comment (adamkdean) [reply src/device/data.ts:1]: P.S. always leave a line after these header comments

File: src/device/node.ts [hunk 9]
```diff
@@ -0,0 +1,33 @@
+import * as data from './data'
```
Comment (adamkdean) [inline src/device/node.ts:1]: Prepend with

```
// Copyright (C) 2021 Edge Network Technologies Limited
// Use of this source code is governed by a GNU GPL-style license
// that can be found in the LICENSE.md file. All rights reserved.
```

File: src/stake/cli.ts [hunk 10]
```diff
@@ -59,27 +39,25 @@ const createAction = (parent: Command, createCmd: Command, network: Network) =>
...
-
-  const encWallet = await readWallet(opts.wallet)
+  const storage = withFile(opts.wallet)
 
   const api = xeWithNetwork(network)
-  const onChainWallet = await api.walletWithNextNonce(encWallet.address)
+  const onChainWallet = await api.walletWithNextNonce(await storage.address())
 
-  // fallback 0 is just for typing - stakeType is checked at top of func, so it should never be used
+  const vars = await xe.vars(network.blockchain.baseURL)
```
Comment (adamkdean) [inline src/stake/cli.ts:47]: Ensure that if this fails for any reason, a message is displayed informing the user that staking is currently unavailable and to try back later

File: src/stake/cli.ts [hunk 11]
```diff
@@ -106,30 +84,32 @@ const createAction = (parent: Command, createCmd: Command, network: Network) =>
...
-    console.log('Your transaction has been submitted and will appear in the explorer shortly.')
-    console.log()
-    console.log(`${network.explorer.baseURL}/transaction/${result.results[0].hash}`)
-  }
+  if (!handleCreateTxResult(network, result)) process.exitCode = 1
 }
 
+const createHelp = (network: Network) => [
+  '\n',
+  'This command will create a stake in the blockchain.\n\n',
```
Comment (adamkdean) [inline src/stake/cli.ts:98]: _on_ the blockchain

File: src/stake/index.ts [hunk 12]
```diff
@@ -0,0 +1,22 @@
+import * as xe from '@edge/xe-utils'
```
Comment (adamkdean) [inline src/stake/index.ts:1]: Prepend with

```
// Copyright (C) 2021 Edge Network Technologies Limited
// Use of this source code is governed by a GNU GPL-style license
// that can be found in the LICENSE.md file. All rights reserved.
```

**Review by adamkdean — CHANGES_REQUESTED** [review]
Generally LGTM. Try to remember to add in the required header comments to new files. Official memo messages should always be Uppercase Every First Character.

File: package.json (hunk 4 above)
Comment (annybs) [reply package.json:43]: Good catch!

**Review by adamkdean — DISMISSED** [review]

**Comment by adamkdean** [comment]
LGTM, do you want to bump the version before you merge this?

**Comment by adamkdean** [comment]
`v1.1.0` 🎉

**Review by adamkdean — APPROVED** [review]
image

**Comment by annybs** [comment]
Yes!

**Pull request merged**